### PR TITLE
Add Nx.conjugate

### PR DESCRIPTION
**Pull request by Benjamin-Philip**
See #693.

### CODE REVIEW

**Comment by Benjamin-Philip** [comment]
@polvalente,

So far I have only added an implementation for `BinaryBackend`. I also haven't written implementations for both EXLA and Torchx before. Are there any pointers/suggestions to keep in mind when writing one?

**Comment by polvalente** [comment]
> 

We can hold off on Torchx and EXLA for now because they don't support Complex numbers yet. I have a WIP branch for Torchx

**Comment by Benjamin-Philip** [comment]
> We can hold off on Torchx and EXLA for now because they don't support Complex numbers yet. I have a WIP branch for Torchx

In that case do we merge, or do we close and wait for complex support in Torchx and EXLA?

File: nx/lib/nx.ex [hunk 1]
```diff
@@ -5192,6 +5192,33 @@ defmodule Nx do
...
+       #Nx.Tensor<
+         c64[2]
+         [1.0-2.0i, 2.0+4.0i]
+       >
+  """
+  @doc type: :element
+  def conjugate(tensor) do
+    tensor = to_tensor(tensor)
+
+    impl!(tensor).conjugate(tensor, tensor)
```
Comment (polvalente) [inline nx/lib/nx.ex:5219]: There's an issue with the output type here.

Try a test/example in which the input is s64 ou f32. I'd expect the output type to be inferred incorrectly (it should be the corresponding complex type).

I think we need a `to_complex` function in Nx.Types, which promotes everything below f32 to c64, f64 -> c128 and keeps complex types as the same.

Then you would have somthing like:
```suggestion
    impl!(tensor).conjugate(%{tensor | type: Nx.Types.to_complex(tensor.type)}, tensor)
```

Comment (polvalente) [reply nx/lib/nx.ex:5219]: Regarding EXLA/Torchx, we can go ahead and merge this after this adjustment :)

Comment (polvalente) [reply nx/lib/nx.ex:5219]: Oh, and regarding the failing torchx tests, there's a list to which you can add the tests in torchx/test/nx_doctest_test.exs

Comment (Benjamin-Philip) [reply nx/lib/nx.ex:5219]: Ah, yes I forgot about it :)

File: nx/lib/nx.ex [hunk 2]
```diff
@@ -5192,6 +5192,38 @@ defmodule Nx do
     end
   end
 
+  @doc """
+  Calculates the complex conjugate of each element in the tensor.
+
+  If $z = a + bi$, $conjugate(z) = z^* = a - bi$
```
Comment (polvalente) [inline nx/lib/nx.ex:5198]:
```suggestion
  If $z = a + bi = r e^\\theta$, $conjugate(z) = z^* = a - bi =  r e^{-\\theta}$
```

File: nx/lib/nx/type.ex [hunk 3]
```diff
@@ -171,6 +171,30 @@ defmodule Nx.Type do
...
+      {:c, 64}
+      iex> Nx.Type.to_complex({:s, 64})
+      {:c, 128}
+      iex> Nx.Type.to_complex({:f, 64})
+      {:c, 128}
+
+  """
+  def to_complex({:c, size}), do: {:c, size}
+  def to_complex({_name, size}) when size <= 32, do: {:c, 64}
+  def to_complex(_type), do: {:c, 128}
```
Comment (polvalente) [inline nx/lib/nx/type.ex:196]: We only need to promote f64 to c128, everything else fits in c64 if it's not already complex

**Review by polvalente — APPROVED** [review]

**Comment by josevalim** [comment]
Should we implement the gradient as we add these new functions?

**Pull request merged**